awsprocesscreds/saml.py

```diff
@@ -216,6 +218,202 @@ def _get_value_of_first_tag(self, root, tag, attr, trait):
+        return self._requests_session.post(
+            url,
+            headers={'Content-Type': 'application/json',
+                     'Accept': 'application/json'},
+            data=json.dumps(body)
+        )
+
+    def process_mfa_sms(self, endpoint, url, statetoken):
+        # Need to trigger the initial code to be sent ...
+        self._password_prompter(("Requesting code to be sent to your phone ..."
+                                 " press RETURN to continue"))
+        self.verify_sms_factor(url, statetoken, "")
+        while True:
+            response = self.get_response(self._MSG_SMS_CODE)
+            if response == "RESEND":
+                response = ""
+            sms_response = self.verify_sms_factor(url, statetoken, response)
+            # If we've just requested a resend, don't check the result
+            # - just loop around to get the next response from the user.
+            if response != "":
+                sms_parsed = json.loads(sms_response.text)
+                if sms_response.status_code == 200:
+                    return self.get_assertion_from_response(endpoint,
+                                                            sms_parsed)
+                elif sms_response.status_code >= 400:
+                    error = sms_parsed["errorCauses"][0]["errorSummary"]
+                    self._password_prompter(("%s\r\n"
+                                             "Press RETURN to continue\r\n")
+                                            % error)
+
+    def display_mfa_choices(self, parsed):
+        index = 1
+        prompt = ""
+        for f in parsed["_embedded"]["factors"]:
+            if f["factorType"] == "token":
+                prompt += "%s: %s token\r\n" % (index, f["provider"])
+            elif f["factorType"] == "token:software:totp":
+                prompt += ("%s: %s authenticator app\r\n"
+                           % (index, f["provider"]))
+            elif f["factorType"] == "sms":
+                prompt += "%s: SMS text message\r\n" % index
+            elif f["factorType"] == "push":
+                prompt += "%s: Push notification\r\n" % index
+            elif f["factorType"] == "question":
+                prompt += "%s: Security question\r\n" % index
+            else:
+                prompt += "%s: %s %s\r\n" % (index,
+                                             f["provider"],
+                                             f["factorType"])
+            index += 1
+        return index, prompt
+
+    def get_mfa_choice(self, parsed):
+        while True:
+            count, prompt = self.display_mfa_choices(parsed)
+            prompt = ("Please choose from the following authentication"
+                      " choices:\r\n") + prompt
+            prompt += ("Enter the number corresponding to your choice "
+                       "or press RETURN to cancel authentication: ")
+            response = self._password_prompter(prompt)
```

Comment:
This change seems to address my first item...

```suggestion
            response = self.get_response(prompt)
```